charms/kfp-persistence/tests/unit/test_operator.py

```diff
@@ -1,262 +1,165 @@
-        ),
-        (
-            # Relation exists with valid data
-            "kfp-api",
-            {
-                "_supported_versions": "- v1",
-                "data": yaml.dump(
-                    {
-                        "service-name": "service-name",
-                        "service-port": "1234",
-                    }
-                ),
-            },
-            {
-                "service-name": "service-name",
-                "service-port": "1234",
-            },
-            does_not_raise(),
-            None,
-        ),
-    ),
-)
-def test_relations_that_provide_data(
-    harness,
-    relation_name,
-    relation_data,
-    expected_returned_data,
-    expected_raises,
-    expected_status,
-):
-    harness.set_leader()
+    # Add relation with data.  This should trigger a charm reconciliation due to relation-changed.
+    add_sdi_relation_to_harness(harness, "kfp-api", data={})
+
+    # Assert
+    assert isinstance(harness.charm.kfp_api_relation.status, BlockedStatus)
+
+
+def test_kfp_api_relation_without_relation(harness):
+    """Test that the kfp-api relation goes Blocked if no relation is established."""
+    # Arrange
     harness.begin()
 
-    other_app = "other-app"
-    other_unit = f"{other_app}/0"
+    # Mock:
+    # * leadership_gate_component_item to be active and executed
+    harness.charm.leadership_gate.get_status = MagicMock(return_value=ActiveStatus())
 
-    if relation_data is not None:
-        rel_id = harness.add_relation(relation_name, other_app)
-        harness.add_relation_unit(rel_id, other_unit)
-        harness.update_relation_data(rel_id, other_app, relation_data)
-
-    with expected_raises as partial_relation_data:
-        interfaces = harness.charm._get_interfaces()
-        data = harness.charm._validate_sdi_interface(interfaces, relation_name)
-    if expected_status is None:
-        assert data == expected_returned_data
-    else:
-        assert partial_relation_data.value.status == expected_status
-
-
-def test_install_with_all_inputs(harness, oci_resource_data):
-    harness.set_leader()
-    model_name = "test_model"
-    harness.set_model_name(model_name)
-
-    # Set up required relations
-    # Future: convert these to fixtures and share with the tests above
-    harness.add_oci_resource(**oci_resource_data)
-
-    # kfp-api relation data
-    os_data = {
-        "_supported_versions": "- v1",
-        "data": yaml.dump(
-            {
-                "service-name": "some-service.some-namespace",
-                "service-port": "1928",
-            }
-        ),
-    }
-    os_rel_id = harness.add_relation("kfp-api", "kfp-api-provider")
-    harness.add_relation_unit(os_rel_id, "kfp-api-provider/0")
-    harness.update_relation_data(os_rel_id, "kfp-api-provider", os_data)
+    # Act
+    harness.charm.on.install.emit()
 
-    harness.begin_with_initial_hooks()
+    # Assert
+    assert isinstance(harness.charm.kfp_api_relation.status, BlockedStatus)
```

Comment:
Are we sure this is Blocked for the expected reason? In my charm, I fell onto a case where even when mocking leadership_gate status as active, the charm was not a leader and couldn't apply K8s resources, thus the K8s componet was blocking. I know this is not exactly your case, but maybe we could be more explicit about the reason the charm's unit is being blocked.

Reply:
Yes. I think, in this case it should be in Blocked.